### ISSUE TITLE
ZFS module crash with   IP: dbuf_find+0x97/0x1a0 [zfs]

### ISSUE THREAD

**Issue by mbaldini1** [openzfs/zfs#6791]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       |   Proxmox VE (based on Debian)
Distribution Version    |    Proxmox 5.1 (based on Debian 9.2)
Linux Kernel                 |   4.13.4-1-pve #1 SMP PVE 4.13.4-25
Architecture                 |   x64
ZFS Version                  |   0.7.2-1
SPL Version                  |  0.7.2-1

### Describe the problem you're observing

I have this server with Proxmox that after update to ZFS 0.7.2  every night have a crash in ZFS module, after the crash the server is still up but disks are unusable. I have / in zfs rpool, virtual machines with kvm have block devices as zfs vols. After the crash,  virtual machines using only memory are working (I have a pfSense VM configured with /var and /tmp in RAM and that still works) but every disk access from proxmox and from virtual machines is impossible, i.e. trying a touch ~/prova will freeze bash forever. The only way to have the disks working again is to hard-reset the server.

My ZFS rpool has a very easy setup, two SATA hdd in RAID-1, I have a NVME Samsung 960 EVO that I used as ZIL and L2ARC but I disabled that after this problem arose, but did not change anything.

```
$ zpool status
  pool: rpool
 state: ONLINE
  scan: scrub repaired 0B in 3h39m with 0 errors on Sun Oct  8 04:03:54 2017
config:

        NAME        STATE     READ WRITE CKSUM
        rpool       ONLINE       0     0     0
          mirror-0  ONLINE       0     0     0
            sdc2    ONLINE       0     0     0
            sdg2    ONLINE       0     0     0

errors: No known data errors


$ zdb
rpool:
    version: 5000
    name: 'rpool'
    state: 0
    txg: 6375513
    pool_guid: 197416030411497508
    errata: 0
    hostid: 2831157250
    hostname: 'pve-hs-main'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 197416030411497508
        children[0]:
            type: 'mirror'
            id: 0
            guid: 720290835225638711
            metaslab_array: 35
            metaslab_shift: 33
            ashift: 12
            asize: 1000190509056
            is_log: 0
            create_txg: 4
            com.delphix:vdev_zap_top: 59
            children[0]:
                type: 'disk'
                id: 0
                guid: 3208937268298614423
                path: '/dev/sdc2'
                whole_disk: 0
                DTL: 1343
                create_txg: 4
                com.delphix:vdev_zap_leaf: 60
            children[1]:
                type: 'disk'
                id: 1
                guid: 12783127680851901291
                path: '/dev/sdg2'
                whole_disk: 0
                DTL: 1342
                create_txg: 4
                com.delphix:vdev_zap_leaf: 61
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data


$ arc_summary
------------------------------------------------------------------------
ZFS Subsystem Report                            Fri Oct 27 08:50:20 2017
ARC Summary: (HEALTHY)
        Memory Throttle Count:                  0

ARC Misc:
        Deleted:                                33.21m
        Mutex Misses:                           11.01k
        Evict Skips:                            11.01k

ARC Size:                               100.07% 6.00    GiB
        Target Size: (Adaptive)         100.00% 6.00    GiB
        Min Size (Hard Limit):          16.67%  1.00    GiB
        Max Size (High Water):          6:1     6.00    GiB

ARC Size Breakdown:
        Recently Used Cache Size:       2.48%   152.36  MiB
        Frequently Used Cache Size:     97.52%  5.86    GiB

ARC Hash Breakdown:
        Elements Max:                           1.46m
        Elements Current:               95.43%  1.39m
        Collisions:                             3.19m
        Chain Max:                              5
        Chains:                                 87.14k

ARC Total accesses:                                     54.47m
        Cache Hit Ratio:                35.82%  19.51m
        Cache Miss Ratio:               64.18%  34.96m
        Actual Hit Ratio:               29.88%  16.27m

        Data Demand Efficiency:         49.40%  31.93m
        Data Prefetch Efficiency:       13.90%  21.39m

        CACHE HITS BY CACHE LIST:
          Anonymously Used:             16.10%  3.14m
          Most Recently Used:           81.50%  15.90m
          Most Frequently Used:         1.91%   372.93k
          Most Recently Used Ghost:     0.42%   81.39k
          Most Frequently Used Ghost:   0.06%   12.19k

        CACHE HITS BY DATA TYPE:
          Demand Data:                  80.84%  15.77m
          Prefetch Data:                15.23%  2.97m
          Demand Metadata:              2.26%   441.31k
          Prefetch Metadata:            1.66%   324.62k

        CACHE MISSES BY DATA TYPE:
          Demand Data:                  46.21%  16.15m
          Prefetch Data:                52.68%  18.42m
          Demand Metadata:              0.42%   146.63k
          Prefetch Metadata:            0.70%   244.03k


DMU Prefetch Efficiency:                                        6.31m
        Hit Ratio:                      32.55%  2.06m
        Miss Ratio:                     67.45%  4.26m



ZFS Tunable:
        zvol_volmode                                      1
        l2arc_headroom                                    2
        dbuf_cache_max_shift                              5
        zfs_free_leak_on_eio                              0
        zfs_free_max_blocks                               100000
        zfs_read_chunk_size                               1048576
        metaslab_preload_enabled                          1
        zfs_dedup_prefetch                                0
        zfs_txg_history                                   0
        zfs_scrub_delay                                   4
        zfs_vdev_async_read_max_active                    3
        zfs_read_history                                  0
        zfs_arc_sys_free                                  0
        l2arc_write_max                                   8388608
        zil_slog_bulk                                     786432
        zfs_dbuf_state_index                              0
        zfs_sync_taskq_batch_pct                          75
        metaslab_debug_unload                             0
        zvol_inhibit_dev                                  0
        zfs_abd_scatter_enabled                           1
        zfs_arc_pc_percent                                0
        zfetch_max_streams                                8
        zfs_recover                                       0
        metaslab_fragmentation_factor_enabled             1
        zfs_deadman_checktime_ms                          5000
        zfs_sync_pass_rewrite                             2
        zfs_object_mutex_size                             64
        zfs_arc_min_prefetch_lifespan                     0
        zfs_arc_meta_prune                                10000
        zfs_read_history_hits                             0
        zfetch_max_distance                               8388608
        l2arc_norw                                        0
        zfs_dirty_data_max_percent                        10
        zfs_per_txg_dirty_frees_percent                   30
        zfs_arc_meta_min                                  0
        metaslabs_per_vdev                                200
        zfs_arc_meta_adjust_restarts                      4096
        spa_load_verify_maxinflight                       10000
        spa_load_verify_metadata                          1
        zfs_multihost_history                             0
        zfs_send_corrupt_data                             0
        zfs_delay_min_dirty_percent                       60
        zfs_vdev_sync_read_max_active                     10
        zfs_dbgmsg_enable                                 0
        zfs_metaslab_segment_weight_enabled               1
        zio_requeue_io_start_cut_in_line                  1
        l2arc_headroom_boost                              200
        zfs_zevent_cols                                   80
        zfs_dmu_offset_next_sync                          0
        spa_config_path                                   /etc/zfs/zpool.cache
        zfs_vdev_cache_size                               0
        dbuf_cache_hiwater_pct                            10
        zfs_multihost_interval                            1000
        zfs_multihost_fail_intervals                      5
        zio_dva_throttle_enabled                          1
        zfs_vdev_sync_write_min_active                    10
        zfs_vdev_scrub_max_active                         2
        ignore_hole_birth                                 1
        zvol_major                                        230
        zil_replay_disable                                0
        zfs_dirty_data_max_max_percent                    25
        zfs_expire_snapshot                               300
        zfs_sync_pass_deferred_free                       2
        spa_asize_inflation                               24
        dmu_object_alloc_chunk_shift                      7
        zfs_vdev_mirror_rotating_seek_offset              1048576
        l2arc_feed_secs                                   1
        zfs_autoimport_disable                            1
        zfs_arc_p_aggressive_disable                      1
        zfs_zevent_len_max                                64
        zfs_arc_meta_limit_percent                        75
        l2arc_noprefetch                                  1
        zfs_vdev_raidz_impl                               [fastest] original scalar sse2 ssse3 avx2
        zfs_arc_meta_limit                                0
        zfs_flags                                         0
        zfs_dirty_data_max_max                            4294967296
        zfs_arc_average_blocksize                         8192
        zfs_vdev_cache_bshift                             16
        zfs_vdev_async_read_min_active                    1
        zfs_arc_dnode_reduce_percent                      10
        zfs_free_bpobj_enabled                            1
        zfs_arc_grow_retry                                0
        zfs_vdev_mirror_rotating_inc                      0
        l2arc_feed_again                                  1
        zfs_vdev_mirror_non_rotating_inc                  0
        zfs_arc_lotsfree_percent                          10
        zfs_zevent_console                                0
        zvol_prefetch_bytes                               131072
        zfs_free_min_time_ms                              1000
        zfs_arc_dnode_limit_percent                       10
        zio_taskq_batch_pct                               75
        dbuf_cache_max_bytes                              104857600
        spa_load_verify_data                              1
        zfs_delete_blocks                                 20480
        zfs_vdev_mirror_non_rotating_seek_inc             1
        zfs_multihost_import_intervals                    10
        zfs_dirty_data_max                                4209932697
        zfs_vdev_async_write_max_active                   10
        zfs_dbgmsg_maxsize                                4194304
        zfs_nocacheflush                                  0
        zfetch_array_rd_sz                                1048576
        zfs_arc_meta_strategy                             1
        zfs_dirty_data_sync                               67108864
        zvol_max_discard_blocks                           16384
        zvol_threads                                      32
        zfs_vdev_async_write_active_max_dirty_percent     60
        zfs_arc_p_dampener_disable                        1
        zfs_txg_timeout                                   5
        metaslab_aliquot                                  524288
        zfs_mdcomp_disable                                0
        zfs_vdev_sync_read_min_active                     10
        zfs_arc_dnode_limit                               0
        dbuf_cache_lowater_pct                            10
        zfs_abd_scatter_max_order                         10
        metaslab_debug_load                               0
        zfs_vdev_aggregation_limit                        131072
        metaslab_lba_weighting_enabled                    1
        zfs_vdev_scheduler                                noop
        zfs_vdev_scrub_min_active                         1
        zfs_no_scrub_io                                   0
        zfs_vdev_cache_max                                16384
        zfs_scan_idle                                     50
        zfs_arc_shrink_shift                              0
        spa_slop_shift                                    5
        zfs_vdev_mirror_rotating_seek_inc                 5
        zfs_deadman_synctime_ms                           1000000
        send_holes_without_birth_time                     1
        metaslab_bias_enabled                             1
        zvol_request_sync                                 0
        zfs_admin_snapshot                                1
        zfs_no_scrub_prefetch                             0
        zfs_metaslab_fragmentation_threshold              70
        zfs_max_recordsize                                1048576
        zfs_arc_min                                       1073741824
        zfs_nopwrite_enabled                              1
        zfs_arc_p_min_shift                               0
        zfs_multilist_num_sublists                        0
        zfs_vdev_queue_depth_pct                          1000
        zfs_mg_fragmentation_threshold                    85
        l2arc_write_boost                                 8388608
        zfs_prefetch_disable                              0
        l2arc_feed_min_ms                                 200
        zio_delay_max                                     30000
        zfs_vdev_write_gap_limit                          4096
        zfs_pd_bytes_max                                  52428800
        zfs_scan_min_time_ms                              1000
        zfs_resilver_min_time_ms                          3000
        zfs_delay_scale                                   500000
        zfs_vdev_async_write_active_min_dirty_percent     30
        zfs_vdev_sync_write_max_active                    10
        zfs_mg_noalloc_threshold                          0
        zfs_deadman_enabled                               1
        zfs_resilver_delay                                2
        zfs_metaslab_switch_threshold                     2
        zfs_arc_max                                       6442450944
        zfs_top_maxinflight                               32
        zfetch_min_sec_reap                               2
        zfs_immediate_write_sz                            32768
        zfs_vdev_async_write_min_active                   2
        zfs_sync_pass_dont_compress                       5
        zfs_vdev_read_gap_limit                           32768
        zfs_compressed_arc_enabled                        1
        zfs_vdev_max_active                               1000

```

### Describe how to reproduce the problem

Hard to say, usually hte problem happens during the night, the only scheduled process running during the night is vzdump (Proxmox own VM backup) but I tried to launch vzdump during day and it went ok. It seems a random crash after some hours of use, in fact after I hard reset the server there are no problems in using it until the night, when it crashes in a random time from 23.00 (11.00 pm) and 4.00 am

### Include any warning/errors/backtraces from the system logs

Before hard-reset the server, I can view dmesg

```
kern  :alert : [Oct26 23:50] BUG: unable to handle kernel paging request at 0000000000c74063
kern  :alert : [  +0.000051] IP: dbuf_find+0x97/0x1a0 [zfs]
kern  :warn  : [  +0.000012] PGD 0
kern  :warn  : [  +0.000000] P4D 0

kern  :warn  : [  +0.000020] Oops: 0000 [#1] SMP
kern  :warn  : [  +0.000009] Modules linked in: tcp_diag inet_diag ip_set xfs ip6table_filter ip6_tables libcrc32c iptable_filter bonding softdog nfnetlink_log nfnetlink ppdev intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel kvm irqbypass crct10dif_pclmul crc32_pclmul ghash_clmulni_intel pcbc aesni_intel aes_x86_64 crypto_simd glue_helper cryptd intel_cstate intel_rapl_perf snd_pcm snd_timer snd soundcore pcspkr i915 hci_uart btbcm serdev btqca btintel bluetooth drm_kms_helper ecdh_generic wmi drm i2c_algo_bit fb_sys_fops syscopyarea sysfillrect sysimgblt parport_pc mei_me parport intel_lpss_acpi mei intel_lpss acpi_als shpchp video tpm_infineon kfifo_buf industrialio acpi_pad mac_hid vhost_net vhost tap ib_iser rdma_cm iw_cm ib_cm sunrpc ib_core iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi
kern  :warn  : [  +0.000180]  ip_tables x_tables autofs4 zfs(PO) zunicode(PO) zavl(PO) icp(PO) zcommon(PO) znvpair(PO) spl(O) btrfs xor raid6_pq uas usb_storage i2c_i801 r8169 8139too 8139cp mii ahci libahci pinctrl_sunrisepoint pinctrl_intel i2c_hid hid
kern  :warn  : [  +0.000057] CPU: 0 PID: 261 Comm: zvol Tainted: P           O    4.13.4-1-pve #1
kern  :warn  : [  +0.000018] Hardware name: Gigabyte Technology Co., Ltd. H170-HD3/H170-HD3-CF, BIOS F20 11/07/2016
kern  :warn  : [  +0.000022] task: ffff9c91476a8000 task.stack: ffffbce286d94000
kern  :warn  : [  +0.000031] RIP: 0010:dbuf_find+0x97/0x1a0 [zfs]
kern  :warn  : [  +0.000013] RSP: 0018:ffffbce286d97b00 EFLAGS: 00010206
kern  :warn  : [  +0.000014] RAX: ffff9c91476a8000 RBX: 0000000000c74063 RCX: 0000000000268986
kern  :warn  : [  +0.000017] RDX: ffff9c91476a8000 RSI: ffffbce294002000 RDI: ffffffffc05a40d0
kern  :warn  : [  +0.000017] RBP: ffffbce286d97b58 R08: 0000000000000001 R09: ffff9c907aae6800
kern  :warn  : [  +0.000017] R10: ffffecf22164d940 R11: 0000000000000000 R12: ffffffffc05a40d0
kern  :warn  : [  +0.000017] R13: 0000000000054010 R14: 0000000000054000 R15: ffff9c91476a8000
kern  :warn  : [  +0.000018] FS:  0000000000000000(0000) GS:ffff9c917ec00000(0000) knlGS:0000000000000000
kern  :warn  : [  +0.000019] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
kern  :warn  : [  +0.000015] CR2: 0000000000c74063 CR3: 000000093a9d2000 CR4: 00000000003426f0
kern  :warn  : [  +0.000017] Call Trace:
kern  :warn  : [  +0.000023]  dbuf_prefetch+0xc4/0x510 [zfs]
kern  :warn  : [  +0.000029]  dmu_zfetch+0x29e/0x4a0 [zfs]
kern  :warn  : [  +0.000027]  dmu_buf_hold_array_by_dnode+0x40b/0x470 [zfs]
kern  :warn  : [  +0.000029]  dmu_read_uio_dnode+0x49/0xf0 [zfs]
kern  :warn  : [  +0.000035]  zvol_read+0xe3/0x280 [zfs]
kern  :warn  : [  +0.000014]  taskq_thread+0x25e/0x460 [spl]
kern  :warn  : [  +0.000013]  ? wake_up_q+0x80/0x80
kern  :warn  : [  +0.000010]  kthread+0x109/0x140
kern  :warn  : [  +0.000616]  ? taskq_thread_should_stop+0x70/0x70 [spl]
kern  :warn  : [  +0.000637]  ? kthread_create_on_node+0x70/0x70
kern  :warn  : [  +0.000631]  ret_from_fork+0x25/0x30
kern  :warn  : [  +0.000611] Code: 00 00 49 89 85 e8 00 55 c0 4c 8b 45 c8 4a 8b 1c fe 49 89 c7 48 85 db 75 12 e9 a6 00 00 00 48 8b 5b 38 48 85 db 0f 84 99 00 00 00 <4c> 3b 03 75 ee 4c 3b 4b 20 75 e8 0f b6 45 c7 3a 43 50 75 df 48
kern  :alert : [  +0.001327] RIP: dbuf_find+0x97/0x1a0 [zfs] RSP: ffffbce286d97b00
kern  :warn  : [  +0.000648] CR2: 0000000000c74063
kern  :warn  : [  +0.000649] ---[ end trace c2004997d5d6b6f4 ]---


```

**Comment by chrisrd** [comment]
tl;dr Looks like a corrupted `db->db_hash_next` in the `dbuf_hash_table`, but how it got there I've no idea. Hopefully someone else can jump in.

...oh, #5742 also concerns bad dbufs, maybe this issue is related? Over there we get an Oops from `__dbuf_hold_impl()` due to a too-large memory copy going out of bounds, perhaps the problem here is a diffent case, where a too-large copy in `__dbuf_hold_impl()` scribbles on our `db->db_hash_next` but doesn't go out of bounds and cause an Oops at that point?

The long version...

The Code line from the Oops gives us this disassembly:
```
$ linux/scripts/decodecode <<END
kern  :warn  : [  +0.000611] Code: 00 00 49 89 85 e8 00 55 c0 4c 8b 45 c8 4a 8b 1c fe 49 89 c7 48 85 db 75 12 e9 a6 00 00 00 48 8b 5b 38 48 85 db 0f 84 99 00 00 00 <4c> 3b 03 75 ee 4c 3b 4b 20 75 e8 0f b6 45 c7 3a 43 50 75 df 48
END

kern :warn : [ +0.000611] Code: 00 00 49 89 85 e8 00 55 c0 4c 8b 45 c8 4a 8b 1c fe 49 89 c7 48 85 db 75 12 e9 a6 00 00 00 48 8b 5b 38 48 85 db 0f 84 99 00 00 00 <4c> 3b 03 75 ee 4c 3b 4b 20 75 e8 0f b6 45 c7 3a 43 50 75 df 48
All code
========
   0:   00 00                   add    %al,(%rax)
   2:   49 89 85 e8 00 55 c0    mov    %rax,-0x3faaff18(%r13)
   9:   4c 8b 45 c8             mov    -0x38(%rbp),%r8
   d:   4a 8b 1c fe             mov    (%rsi,%r15,8),%rbx
  11:   49 89 c7                mov    %rax,%r15
  14:   48 85 db                test   %rbx,%rbx
  17:   75 12                   jne    0x2b
  19:   e9 a6 00 00 00          jmpq   0xc4
  1e:   48 8b 5b 38             mov    0x38(%rbx),%rbx
  22:   48 85 db                test   %rbx,%rbx
  25:   0f 84 99 00 00 00       je     0xc4
  2b:*  4c 3b 03                cmp    (%rbx),%r8               <-- trapping instruction
  2e:   75 ee                   jne    0x1e
  30:   4c 3b 4b 20             cmp    0x20(%rbx),%r9
  34:   75 e8                   jne    0x1e
  36:   0f b6 45 c7             movzbl -0x39(%rbp),%eax
  3a:   3a 43 50                cmp    0x50(%rbx),%al
  3d:   75 df                   jne    0x1e
  3f:   48                      rex.W

Code starting with the faulting instruction
===========================================
   0:   4c 3b 03                cmp    (%rbx),%r8
   3:   75 ee                   jne    0xfffffffffffffff3
   5:   4c 3b 4b 20             cmp    0x20(%rbx),%r9
   9:   75 e8                   jne    0xfffffffffffffff3
   b:   0f b6 45 c7             movzbl -0x39(%rbp),%eax
   f:   3a 43 50                cmp    0x50(%rbx),%al
  12:   75 df                   jne    0xfffffffffffffff3
  14:   48                      rex.W
```


The source gives us:
```
#define DBUF_EQUAL(dbuf, os, obj, level, blkid)         \
        ((dbuf)->db.db_object == (obj) &&               \
        (dbuf)->db_objset == (os) &&                    \
        (dbuf)->db_level == (level) &&                  \
        (dbuf)->db_blkid == (blkid))

dmu_buf_impl_t * 
dbuf_find(objset_t *os, uint64_t obj, uint8_t level, uint64_t blkid)
{        
        dbuf_hash_table_t *h = &dbuf_hash_table;
        uint64_t hv;
        uint64_t idx;
        dmu_buf_impl_t *db;

        hv = dbuf_hash(os, obj, level, blkid);
        idx = hv & h->hash_table_mask;
 
        mutex_enter(DBUF_HASH_MUTEX(h, idx));
        for (db = h->hash_table[idx]; db != NULL; db = db->db_hash_next) {
                if (DBUF_EQUAL(db, os, obj, level, blkid)) {
                ...
                }
        }
        ...
}
```

Looking at `dmu_buf_impl_t *db;`, a `struct dmu_buf_impl_t` has these (hex) offsets:
```
typedef struct dmu_buf_impl {
        dmu_buf_t db == struct dmu_buf {
0           uint64_t db_object;
8           uint64_t db_offset;
10          uint64_t db_size;
18          void *db_data;
        }
20      struct objset *db_objset;
28      struct dnode_handle *db_dnode_handle;
30      struct dmu_buf_impl *db_parent;
38      struct dmu_buf_impl *db_hash_next;
40      uint64_t db_blkid;
48      blkptr_t *db_blkptr;
50      uint8_t db_level;
...
}
```

The disassembly at `1e: mov 0x38(%rbx),%rbx` looks like source `db = db->db_hash_next` (the `0x38(%rbx)` matches the offset of `db->db_hash_next`).

The faulting instruction, `2b: cmp (%rbx),%r8`, is the first comparison in `DBUF_EQUAL`, i.e. `(dbuf)->db.db_object == (obj)` (the 0 offset on `(%rbx)` matches the 0 offset of `db->db.db_object`). From this we can deduce `%rbx` is `db` in the source, and %r8 is `obj`, which our Oops tells us is 1.

The faulting address is given in `%CR2`, i.e. `0xc74063`, which matches `%RBX` from the Oops. I.e. our `%rbx` (== `db`) is out of range (and `0xc74063` certainly doesn't look like a kernel address). And we can see that `db` was set by `db = db->db_hash_next`, i.e. the previous `db->db_hash_next` was out of range.

A little further down in the disassembly, `30: cmp 0x20(%rbx),%r9` translates to source `(dbuf)->db_objset == (os)`. Which means our `os` is in `%r9`, i.e. `0xffff9c907aae6800`, which doesn't tell us much, but at least it looks like a proper kernel address.

Then `3a: cmp 0x50(%rbx),%al` translates to `(dbuf)->db_level == (level)`, so `level` is the lowest byte of `%rax`, i.e. 0.

I.e. we were called like:
```
dbuf_find(objset_t *os=0xffff9c907aae6800, uint64_t obj=1, uint8_t level=0, uint64_t blkid=???)
```

Unfortunately that's about as far as I can take it.

**Comment by chrisrd** [comment]
@mbaldini1If you're comfortable patching and compiling your own ZoL, you could try @dweeezil's patch in #6797 to see if this also solves your problem.
